ext/openssl/openssl.c

```diff
@@ -2914,29 +3015,41 @@ static X509_REQ * php_openssl_csr_from_zval(zval * val, int makeresource, zend_r
 
 	return csr;
 }
-/* }}} */
+
+static X509_REQ *php_openssl_csr_from_param(zend_object *csr_obj, zend_string *csr_str)
+{
+	if (csr_obj) {
+		return x509_request_from_obj(csr_obj)->csr;
+	}
+
+	ZEND_ASSERT(csr_str);
+
+	return php_openssl_csr_from_str(csr_str);
+}
 
 /* {{{ Exports a CSR to file */
 PHP_FUNCTION(openssl_csr_export_to_file)
 {
-	X509_REQ * csr;
-	zval * zcsr = NULL;
+	X509_REQ *csr;
+	zend_object *csr_obj;
+	zend_string *csr_str;
 	zend_bool notext = 1;
 	char * filename = NULL;
 	size_t filename_len;
 	BIO * bio_out;
-	zend_resource *csr_resource;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "rp|b", &zcsr, &filename, &filename_len, &notext) == FAILURE) {
```

Comment:
It's very weird... The CSR is a resource, however `php_openssl_csr_from_zval()` treated it as if it was `resource|string`. Seems like a bug or dead code :) My guess is that it's a bug since freeing of `csr` is based on the assumption that it can come from a string.

Reply:
I think `php_openssl_csr_from_zval` was just used to verify that it's a correct resource type. String doesn't make much sense here because you could just write it directly to file. It would add some extra check but not sure if it's really worth it but if you want to do it for consistency, then fine.

Reply:
What is our conclusion? I wouldn't mind getting rid of the union types here, if `string` doesn't make sense.

Reply:
I think it's fine to leave it as union. It can be kind of verification that the supplied string is really valid CSR PEM.